data_structures/src/mainnet_validations.rs

```diff
@@ -28,6 +30,257 @@ pub const SECOND_HARD_FORK: Epoch = 376320;
+    pub wip_activation: HashMap<String, Epoch>,
+}
+
+impl TapiEngine {
+    pub fn update_bit_counter(
+        &mut self,
+        v: u32,
+        epoch_to_update: Epoch,
+        block_epoch: Epoch,
+        avoid_wip_list: &HashSet<String>,
+    ) {
+        // In case of empty epochs, they would be considered as blocks with tapi version to 0
+        // In order to not update bit counter from old blocks where the block version was not used,
+        // the first time (bit_tapi_counter.last_epoch == 0) would be skipped in this conditional branch
+        if self.bit_tapi_counter.last_epoch != 0
+            && epoch_to_update > self.bit_tapi_counter.last_epoch + 1
+        {
+            let init = self.bit_tapi_counter.last_epoch + 1;
+            let end = epoch_to_update;
+            for i in init..end {
+                self.update_bit_counter(0, i, block_epoch, avoid_wip_list);
+            }
+        }
+        for n in 0..self.bit_tapi_counter.len() {
+            if let Some(mut bit_counter) = self.bit_tapi_counter.get_mut(n, &epoch_to_update) {
+                if !self.wip_activation.contains_key(&bit_counter.wip)
+                    && !avoid_wip_list.contains(&bit_counter.wip)
+                {
+                    if is_bit_n_activated(v, n) {
+                        bit_counter.votes += 1;
+                    }
+                    if (epoch_to_update - bit_counter.init) % bit_counter.period == 0 {
+                        if (bit_counter.votes * 100) / bit_counter.period >= 80 {
+                            // An offset of 21 is added to ensure that the activation of the WIP is
+                            // achieved with consolidated blocks
+                            self.wip_activation
+                                .insert(bit_counter.wip.clone(), block_epoch + 21);
+                        }
+                        bit_counter.votes = 0;
+                    }
+                }
+            }
+        }
+        self.bit_tapi_counter.last_epoch = epoch_to_update;
+    }
+
+    pub fn initialize_wip_information(
+        &mut self,
+        environment: Environment,
+    ) -> (Epoch, HashSet<String>) {
+        let mut voting_wips = vec![None; 32];
+
+        match environment {
+            Environment::Mainnet => {
+                // Hardcoded information about WIPs
+                self.wip_activation
+                    .insert("WIP0008".to_string(), FIRST_HARD_FORK);
+                self.wip_activation
+                    .insert("WIP0009-0011-0012".to_string(), SECOND_HARD_FORK);
+                self.wip_activation
+                    .insert("THIRD_HARD_FORK".to_string(), THIRD_HARD_FORK);
+
+                // Hardcoded information about WIPs in vote processing
+                let bit = 0;
+                let wip_0014 = BitVotesCounter {
+                    votes: 0,
+                    period: 26880,
+                    wip: "WIP0014".to_string(),
+                    init: 500000,
```

Comment:
Should we include a TODO here, to remember to modify the date?

Reply:
Ok! Also, should it be WIP0014, WIP0016, or both?

Reply:
I think both

Reply:
Done, renamed to "WIP0014-0016" and included "oppose_wip0016" flag in tapi config.